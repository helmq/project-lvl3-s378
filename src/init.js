import { watch } from 'melanke-watchjs';
import { isURL } from 'validator';
import axios from 'axios';
import _ from 'lodash';
import $ from 'jquery';
import parseRSS from './rss';

const proxy = 'http://cors-anywhere.herokuapp.com/';

const updateArticles = (state) => {
  const urls = Array.from(state.urls);
  const promises = urls.map(url => axios.get(`${proxy}${url}`));
  Promise.all(promises).then((allData) => {
    const fetchedArticles = _.flatten(allData.map(parseRSS).map(content => content.articles));
    const currentArticles = state.articles;
    const newArticles = _.differenceBy(fetchedArticles, currentArticles, 'link');
    if (newArticles.length > 0) {
      state.addArticles(newArticles);
    }
  }).catch((e) => {
    console.log(e.message);
  }).finally(() => {
    setTimeout(updateArticles, 5000, state);
  });
};
const updateFeed = (state) => {
  if (!state.isUrlValid || state.url === '') {
    state.setErrorMessage('Please enter correct URL');
    return;
  }
  if (state.isRequestSubmitting) {
    state.setErrorMessage('Please wait');
    return;
  }
  if (state.urls.has(state.url)) {
    state.setErrorMessage('URL is already exists');
    return;
  }
  state.setRequestStatusSubmitting();
  axios.get(`${proxy}${state.url}`).then((content) => {
    try {
      const { title, description, articles } = parseRSS(content);
      state.addChannel({ title, description });
      state.addArticles(articles);
      state.setRequestStatusDone();
      state.addUrl(state.url);
      if (!state.checkingUpdates) {
        updateArticles(state);
        state.toggleCheckingUpdates();
      }
    } catch (e) {
      state.setErrorMessage(e.message);
      state.setRequestStatusDone();
    }
  }).catch((e) => {
    state.setErrorMessage(e.message);
    state.setRequestStatusDone();
  });
};

const onChangeUrl = (state) => {
  const { url } = state;
  const isUrlValid = isURL(url) || url === '';
  if (isUrlValid) {
    state.setErrorMessage('');
    state.urlIsValid();
  } else {
    state.urlIsInvalid();
  }
  state.setUrl(url);
};

export default (state) => {
  const form = document.getElementById('rss-form');
  const input = form.elements.url;
  const channelsFeed = document.getElementById('channels-feed');
  const articlesFeed = document.getElementById('articles-feed');
  const modal = document.getElementById('description-modal');

  watch(state, ['isUrlValid', 'errorMessage'], () => {
    if (state.isUrlValid || state.errorMessage === '') {
      input.classList.remove('is-invalid');
    }
    if (!state.isUrlValid) {
      input.classList.add('is-invalid');
    }

    const currentFeedbackEl = input.parentNode.querySelector('.invalid-feedback');
    if (currentFeedbackEl) {
      input.parentNode.removeChild(currentFeedbackEl);
    }
    if (state.errorMessage !== '') {
      const feedbackEl = document.createElement('div');
      feedbackEl.classList.add('invalid-feedback');
      feedbackEl.textContent = state.errorMessage;
      input.classList.add('is-invalid');
      input.parentNode.appendChild(feedbackEl);
    }
  });
  watch(state, ['channels', 'articles'], () => {
    const { articles, channels } = state;
    const channelsHTML = channels.map(({ title, description }) => `
      <div>
        <h4>${title}</h4>
        <p>${description}</p>
      </div>`);
    const articlesHTML = articles.map(({ title, id, link }) => `
        <div class="row">
          <div class="col-8">
            <a href="${link}" target="_blank">${title}</a>
          </div>
          <div class="col">
            <button type="button" class="btn btn-primary" data-toggle="modal" data-id="${id}">Show description</button>
          </div>
        </div>
      `);
    channelsFeed.innerHTML = channelsHTML.join('');
    articlesFeed.innerHTML = articlesHTML.join('<hr>');
    $('button[data-toggle=modal]').click((e) => {
      const { id } = e.target.dataset;
      const { description, title } = _.find(articles, a => a.id === id);
      const modalBody = modal.querySelector('.modal-body');
      const modalTitle = modal.querySelector('.modal-title');
      modalBody.innerHTML = `<p>${description}</p>`;
      modalTitle.innerHTML = title;
      $(modal).modal('show');
    });
  });
  watch(state, 'channels', () => {
    form.reset();
  });

  input.addEventListener('input', (e) => {
    state.setUrl(e.target.value);
    onChangeUrl(state);
  });
  form.addEventListener('submit', (e) => {
    e.preventDefault();
    updateFeed(state);
    return false; // prevent submit for Safari
  });
};
