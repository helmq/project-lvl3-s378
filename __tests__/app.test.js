import { promises as fs } from 'fs';
import path from 'path';
import { html } from 'js-beautify';
import timer from 'timer-promise';
import init from '../src/init';

beforeEach(async () => {
  const pathToHtml = path.resolve(__dirname, '__fixtures__/index.html');
  const innerHTML = await fs.readFile(pathToHtml, 'utf8');
  document.body.innerHTML = innerHTML;
});

const getTree = () => html(document.body.innerHTML);

test('test incorrect url', async () => {
  init();
  const form = document.getElementById('rss-form');
  const input = form.elements.url;
  const button = form.elements.submit;
  input.value = 'test';
  input.dispatchEvent(new Event('input'));
  button.click();
  await timer.start('test', 100);
  expect(getTree()).toMatchSnapshot();
});
