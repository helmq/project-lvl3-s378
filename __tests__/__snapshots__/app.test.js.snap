// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test incorrect url 1`] = `
"<title>
    RSS reader
</title>



<div class=\\"modal\\" tabindex=\\"-1\\" role=\\"dialog\\" id=\\"description-modal\\">
    <div class=\\"modal-dialog\\" role=\\"document\\">
        <div class=\\"modal-content\\">
            <div class=\\"modal-header\\">
                <h5 class=\\"modal-title\\"></h5>
                <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\" aria-label=\\"Close\\">
                    <span aria-hidden=\\"true\\">Ã—</span>
                </button>
            </div>
            <div class=\\"modal-body\\">
            </div>
            <div class=\\"modal-footer\\">
                <button type=\\"button\\" class=\\"btn btn-secondary\\" data-dismiss=\\"modal\\">Close</button>
            </div>
        </div>
    </div>
</div>
<div class=\\"jumbotron\\">
    <div class=\\"container\\">
        <h1 class=\\"display-4\\">
            RSS reader
        </h1>
        <p class=\\"lead\\">This is a simple RSS reader.</p>
        <hr class=\\"my-4\\">
        <form id=\\"rss-form\\">
            <div class=\\"form-group row\\">
                <div class=\\"col-5\\">
                    <input class=\\"form-control is-invalid\\" type=\\"text\\" placeholder=\\"Name\\" name=\\"url\\">
                    <div class=\\"invalid-feedback\\">Please enter correct URL</div>
                </div>
                <div class=\\"col\\">
                    <button type=\\"submit\\" class=\\"btn btn-primary\\" name=\\"submit\\">Submit</button>
                </div>
            </div>
        </form>
    </div>
</div>
<div class=\\"container\\">
    <div class=\\"row\\">
        <div class=\\"col-6\\">
            <h3>Channel list</h3>
            <div id=\\"channels-feed\\">Nothing added yet.</div>
        </div>
        <div class=\\"col-6\\">
            <h2>Feed</h2>
            <div id=\\"articles-feed\\"></div>
        </div>
    </div>
</div>"
`;
